popupmnu.c: Fix popup placement when preview window is below

rpigott opened the pull request:
The popup menu doesn't adjust correctly for new preview window when the preview window is below the current window like with 'splitbelow' set. Appears to be from a typo, and vim8 has the correct behavior. This PR just changes it to match vim.

Heres an album which shows the bug I'm talking about. https://imgur.com/a/gBwRF0c .

Commit:
popupmnu.c: Fix popup placement when preview window is below

Commit:
vim-patch:8.0.1161

justinmk commented:
Thanks, this is appreciated. Can you fix the commit message so there's no whitespace in `vim-patch:8.0.1161`. Would also help to fix the lint errors (in a separate commit titled "lint"), run `make lint` to see them.

Commit:
lint

RPigott commented:
So I tried to match the behavior of vim, but do you guys think we could deviate a bit here? There are two behaviors I don't like.

Vim allows the popup to draw over the status line of the preview window.

Vim retains space between the pum_row and the preview window, e.g. if the popup is inverted and the top row is at line 9, and a new preview window appears above and is 10 lines tall, the top of the popup will be squished by 10 lines instead of the 1 line necessary to not draw over it.

Can I put these changes here? It seems better all around.

janlazo commented:
@RPigott Do it in a separate PR so that it's safer to merge for the next point release. Rebase on top of this PR if you have to. Neovim hasn't merged all of the patches that affect `popupmnu.c` so, worst case, your deviation may get removed to make tests pass. It happened for `test_vimscript.vim` at https://github.com/neovim/neovim/pull/8702.

bfredl commented:
>  Neovim hasn't merged all of the patches that affect popupmnu.c so, worst case, your deviation may get removed to make tests pass. 

Though for visual features, it is better to make our own nvim-style screen tests, rather than copy vim's more quirky ways of asserting screen state. Our own tests can then assert the improvements, so we can check they are preserved across merging of vim patches. Also if it is not too much work, the improvement could be submitted to upstream vim as well.

justinmk commented on src/nvim/screen.c:
```diff
@@ -7102,11 +7102,13 @@ void screen_resize(int width, int height)
-        } else
+          ins_compl_show_pum();
+        } else {
           update_screen(NOT_VALID);
```
in the [vim patch](https://github.com/vim/vim/commit/a5e6621aadadf78c7b344e93a4b328788076f14c?w=1) this `else` was removed, so that `update_screen(NOT_VALID)` is always called. (The vim commit message hints about this: "also when a popup menu is visible").

Commit:
fixup patch 8.0.1161

Commit:
[WIP] popupmenu placement test

RPigott commented:
Okay thanks, @justinmk. I've added your corrections.

I've decided to provide some appropriate tests in 'test/functional' as well. I noticed that someone has since included 'test/functional/ui/popupmenu_spec.lua' in upstream master so I rebased and included it there. The tests are not working at the moment.

RPigott commented:
For some reason, I can't get the test to match the behavior I observe with `nvim -u NORC`. I could be wrong, but it appears that the preview window draws OVER the popup menu in the test screen, but that's not what I observe when I shrink my terminal to the given size and try it myself with `nvim -u NORC`  (or any other usage of neovim, for that matter) where the popup menu is on top of everything. Other than that, the titular bug seems to exist, where the menu is drawn in the wrong place, I just can't get the test to have the same incorrect behavior I observe. I'll push again when I can figure this out.

Commit:
Added test to test for popup misplacement behavior with preview window

Commit:
Add all 4 test cases with pvwin above and below.

RPigott commented:
I've made 4 tests that cover the cases I checked when writing the fix. Where the popup is above (inverted) / below the cursor line and the preview window is above / below the cursor window.

The behavior for normal popup and pvwin above is unchanged, but it probably deserves a test so its got one.

The behavior for normal popup and pvwin below is fixed to match vim, per the original purpose of the PR.

The behavior for inverted popup and pvwin above is unchanged but awkward, and I think I'll submit a PR to vim to have it changed there and then here.

The behavior for inverted popup and pvwin below is fixed to match vim.

RPigott commented on src/nvim/testdir/test_popup.vim:
```diff
@@ -665,4 +667,30 @@ func Test_complete_CTRLN_startofbuffer()
 endfunc
 
+func Test_popup_and_window_resize()
+  if !has('terminal') || has('gui_running')
```
I've added some screen tests in `test/functional/ui/`

Commit:
Change 'pvwin' to 'preview-window' in test names.

justinmk commented:
Merged after squashing some of the intermediate commits.
